crates/proto/src/runtime_provider.rs

```diff
@@ -0,0 +1,67 @@
+//! Runtime abstraction component for DNS clients
+use crate::{error::ProtoError, tcp::DnsTcpStream, udp::UdpSocket, Time};
+use std::{future::Future, io, net::SocketAddr};
+
+/// RuntimeProvider defines which async runtime that handles IO and timers.
+#[async_trait::async_trait]
+pub trait RuntimeProvider: Clone + 'static + Send + Sync + Unpin {
+    /// Time implementation used for this type
+    type Time: Time + Unpin + Send;
+    /// Type of socket that would be bound by the trait implementation. E.g. for tokio, it would be
+    /// `tokio::net::UdpSocket`.
+    type UdpSocket: UdpSocket + Send + 'static;
+
+    /// Socket type that is returned after a successful connection.
+    type TcpConnection: DnsTcpStream;
+
+    /// Bind an UDP socket to the given socket address.
+    async fn bind_udp(&self, addr: SocketAddr) -> io::Result<Self::UdpSocket>;
```

Comment:
I think it would be better to have these iterfaces consistent, tcp and udp. If this being `async_triat` causes problems, then let's do that. I think making them both `&self` makes sense to me.